### DISCUSSION TITLE
Normalized grades/ranks

### DISCUSSION THREAD

**Discussion by MitsukaiDokuro** [ppy/osu#13131] [Ideas]
**Describe the new feature:** Directly compute score grades from accuracy, based on a natural logarithmic scale. It makes a lot more sense and just feels intuitive.

**Proposal designs of the feature:** 

- Compute `grade = -ln(1 - accuracy)` (reverse would be `accuracy = 1 - exp(-grade)`) (accuracy running from 0.0 to 1.0)
- Look up grade name/icon from a table like this (top to bottom):
`grade <= 1` → C rank
`grade <= 2` → B rank
`grade <= 3` → A rank
`grade < ∞` → S rank
`grade = ∞` → SS rank
- If played with visually impaired (HD, FL mods) apply platinum icons instead
- Optionally, subdivide into ranking system into C-, C0, C+, B-, ... or increase "shininess" of icon or something. Lots of options.

**Comment by iSimpFor** [comment]
so if i calculated every thing correctly,
C would need 63.21% accuracy
B would need 86.47% accuracy
A would need 95.02% accuracy
idk how to calculate S, infinity is weird
SS is still perfect

is this correct?

**Comment by Firmatorenio** [comment]
After fiddling around with numbers, this suggests:
C - up to 63.21%;
B - 63.21..86.47%;
A - 86.47..95.02%;
S - over 95.02% but not 100%.
It completely shuns the D rank (which is a valid rank in my opinion).

Personally, I think the current system (D<70<C<80<B<90<A<95<S) is perfectly fair, (actually it feels a lot fairer coming from taiko where you needed to get over ~94% accuracy to get an A on Stable).

In my mind, S means 'amazing', A means 'great', B means 'decent', C means 'OK' and D means 'barely passed'. The semantics can be argued, but this system suggests a 65% accuracy play is worth a B rank, which, if I'm honest, is kind of outrageous.

**Comment by iSimpFor** [comment]
>     * Look up grade name/icon from a table like this (top to bottom):
>       `grade <= 1` → C rank
>       `grade <= 2` → B rank
>       `grade <= 3` → A rank
>       `grade < ∞` → S rank
>       `grade = ∞` → SS rank
maybe instead of the list above, we adjust to this
>     * Look up grade name/icon from a table like this (top to bottom):
>       `grade <= 1.5` → C rank
>       `grade <= 2.25` → B rank
>       `grade <= 3` → A rank
>       `grade < ∞` → S rank
>       `grade = ∞` → SS rank
it would make getting a B more realistic than getting what's considered the current D

**Comment by Firmatorenio** [comment]
I think that pretty much defeats the original idea of natural logarithms - to an outside observer you're taking rather arbitrary accuracy numbers to define the rank borders (in this case, C<77.69<B<89.46<A<95.02<S), then you try to explain this system... which again uses arbitrary (albeit nicer) numbers behind the scenes (1.5, 2.25 and 3). There's no need for such complexity, in my opinion. The current system is very straightforward and pretty fair.

**Comment by abraker95** [comment]
Yea as @Firmatorenio said, this adds an extra layer of unnecessary complexity.

**Comment by peppy** [comment]
I'm fine with making changes, but I also disagree with making them such precise numbers. The cutoffs should be multiples of 5 (or at very least whole int vals) because players don't want to have to remember fractions.

**Comment by MitsukaiDokuro** [comment]
@igontusuwu The calculation is correct. If you are going to modify the grade name/icon lookup table. I'm fine with your proposed modification to the lookup table - if you insert another point for "D rank" at 0.75 (to close the gap). For future readers : It's important to make the numeric values of the entries equally spaced to each other, including the implicit entry for 0.

@Firmatorenio Well, it's subjective which system you would prefer. This isn't even complex maths, in my opinion.

@peppy Well, the whole point of the suggestion was to make grading more precise and intuitive without relying on arbitrary numbers (with exception of the point from where on a score is rated as 'S' and the number of grades). Rounding to multiples of 5 would be counter-productive. Though whole integer rounding would be fine, I guess.

**Comment by iSimpFor** [comment]
if this idea is still on, i would like to make a adjustment to the grades:

currently suggested:
grade <= 1 → C rank
grade <= 2 → B rank
grade <= 3 → A rank
grade < ∞ → S rank
grade = ∞ → SS rank

adjust to:
grade <= 1 → D rank (under ~63.2%)
grade <= 2 → C rank (under ~86.5%)
grade <= 3 → B rank (under ~95%)
grade <= 4 → A rank (under ~98.2%)
(optional) grade <= 5 → A rank, shift everything down one letter grade if used (under ~99.3%)
(optional) grade <= 6 → A rank, shift everything down one letter grade if used (under ~99.8%)
grade < ∞ → S rank (under 100%)
grade = ∞ → SS rank (100%)

i think this fits better because i think the S should really show that you can play well while shifting the lower grades closer to where they feel like they should be

said values can be rounded if needed

**Reply by uzervlad** [reply]
<95% being a B rank feels a bit extreme

**Reply by iSimpFor** [reply]
i think it just happens to be like that with the natural logarithm, no other way to set the values without using decimals which peppy said that the players wouldn't find nice

**Comment by uzervlad** [comment]
@Moltanicaa 's suggestion is:

- **D** is below 70%
- **C** is 70-84%
- **B** is 84-94%
  - *Optionally, we could reduce the range to about 90-94%, in case there are too many 50s and misses (no certain value yet)*
- **A** is 94-98% with no special rules
- **S** is above 98% with amount of misses being less than 1% of amount of hitobjects
- **SS** is a perfect play
